File: sdk/react-native/dazzle-react-native/src/primitives/list.ts
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0

import { asLong, asBulkArray, asString, dazzleCommand } from '../ffi/command';

export class ListKey {
  constructor(public readonly key: string) {}

  async push(...values: string[]): Promise<number> {
    if (values.length === 0) return 0;
    const r = await dazzleCommand(['RPUSH', this.key, ...values]);
    return asLong(r) ?? 0;
  }

  async pop(): Promise<string | null> {
    const r = await dazzleCommand(['LPOP', this.key]);
    return asString(r);
  }

  async range(start: number, stop: number): Promise<string[]> {
    const r = await dazzleCommand(
        ['LRANGE', this.key, String(start), String(stop)]);
    const arr = asBulkArray(r);
    return arr ? arr.filter((x): x is string => x !== null) : [];
  }

  async length(): Promise<number> {
    const r = await dazzleCommand(['LLEN', this.key]);
    return asLong(r) ?? 0;
  }
}

File: sdk/flutter/dazzle_flutter/lib/src/primitives/list.dart
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0

import '../ffi/command.dart';

class ListKey {
  ListKey(this.key);
  final String key;

  int push(Iterable<String> values, {bool atHead = false}) {
    if (values.isEmpty) return 0;
    final cmd = atHead ? 'LPUSH' : 'RPUSH';
    return dazzleCommand([cmd, key, ...values]).asLongOrNull ?? 0;
  }

  String? pop({bool atHead = true}) {
    final cmd = atHead ? 'LPOP' : 'RPOP';
    return dazzleCommand([cmd, key]).asStringOrNull;
  }

  int size() => dazzleCommand(['LLEN', key]).asLongOrNull ?? 0;

  List<String> range(int start, int stop) {
    final r = dazzleCommand(['LRANGE', key, '$start', '$stop']);
    return r.asBulkArrayOrNull?.whereType<String>().toList() ?? const [];
  }

  bool deleteKey() => (dazzleCommand(['DEL', key]).asLongOrNull ?? 0) > 0;
}
